skimage/segmentation/morphsnakes.py

```diff
@@ -0,0 +1,440 @@
+                         "(should be 2 or 3)")
+
+    erosions = []
+    for P_i in P:
+        erosions.append(ndi.binary_erosion(u, P_i))
+
+    return np.array(erosions, dtype=np.int8).max(0)
+
+
+def inf_sup(u):
+    """IS operator."""
+
+    if np.ndim(u) == 2:
+        P = _P2
+    elif np.ndim(u) == 3:
+        P = _P3
+    else:
+        raise ValueError("u has an invalid number of dimensions "
+                         "(should be 2 or 3)")
+
+    dilations = []
+    for P_i in P:
+        dilations.append(ndi.binary_dilation(u, P_i))
+
+    return np.array(dilations, dtype=np.int8).min(0)
+
+
+curvop = _fcycle([lambda u: sup_inf(inf_sup(u)),   # SIoIS
+                  lambda u: inf_sup(sup_inf(u))])  # ISoSI
+
+
+def _check_input(image, init_level_set):
+    """Check that shapes of `image` and `init_level_set` match."""
+    if len(image.shape) not in [2, 3]:
+        raise ValueError("Input image is not a 2D or 3D array.")
+
+    if len(image.shape) != len(init_level_set.shape):
+        raise ValueError("The dimensions of the initial level set do not "
+                         "match the dimensions of the image.")
+
+
+def _find_threshold(image, percentile=20):
+    """Find a proper threshold for `morph_gac` using the given percentile."""
+    return np.percentile(image, percentile)
+
+
+def _init_level_set(init_level_set, image_shape):
+    """Auxiliary function for initializing level sets with a string.
+
+    If `init_level_set` is not a string, it is returned as is.
+    """
+    if isinstance(init_level_set, str):
+        if init_level_set == 'checkerboard':
+            res = checkerboard_level_set(image_shape)
+        elif init_level_set == 'circle':
+            res = circle_level_set(image_shape)
+        else:
+            raise ValueError("`init_level_set` not in "
+                             "['checkerboard', 'circle']")
+    else:
+        res = init_level_set
+    return res
+
+
+def circle_level_set(image_shape, center=None, radius=None):
+    """Create a circle level set with binary values.
+
+    Parameters
+    ----------
+    image_shape : tuple of positive integers
+        Shape of the image
+    center : tuple of positive integers, optional
+        Coordinates of the center of the circle given in (row, column). If not
+        given, it defaults to the center of the image.
+    radius : float, optional
+        Radius of the circle. If not given, it is set to the 75% of the
+        smallest image dimension.
+
+    Returns
+    -------
+    out : array with shape `image_shape`
+        Binary level set of the circle with the given `radius` and `center`.
+
+    See also
+    --------
+    checkerboard_level_set
+    """
+
+    if center is None:
+        center = tuple(i // 2 for i in image_shape)
+
+    if radius is None:
+        radius = min(image_shape) * 3.0 / 8.0
+
+    grid = np.mgrid[[slice(i) for i in image_shape]]
+    grid = (grid.T - center).T
```

Comment:
Is this correct? `grid.T - center` will subtract the row center from the column coordinates and the column center from the row coordinates?

An alternative:

```python
grid -= np.reshape(center, (ndim,) + (1,) * ndim)
```

or:

```python
grid = np.mgrid[[slice(-center[i], image_shape[i] - center[i]) for i in range(ndim)]]
```

Reply:
`grid.T - center` is tricky, but it is surprisingly correct (I had to check it).

Note that `grid[:. i, j]` contains the coordinate in format (row, column). Also, `grid[:, i, j] == grid.T[j, i, :]` for all `i`, `j`. That means that `grid.T[j, i, :]` also contains the coordinate in format (row, column). For example,

```python
In [25]: grid = np.mgrid[:3, :4]

In [26]: grid[:, 2, 3]
array([2, 3])

In [27]: grid.T[3, 2, :]
array([2, 3])
```

Therefore, subtracting `center` from `grid.T` is correct.

In any case, if you prefer the `reshape` approach for being less confusing, I'm open to change it.

Reply:
Oh of course! I confused transposing with flipping. Personally I find the `-=` approach to be the most readable, but change only if you agree... I can see how the reshape argument is itself confusing.

Reply:
Thanks for the long exposition. =)